### PR TITLE
:sparkles: Add kai configurables to extension config

### PR DESCRIPTION
**Pull request by djzager**
This PR adds the kai configuration items to our extension's config **AND** attempts to standardize the way we get and update our configuration items.

### CODE REVIEW

File: vscode/src/client/analyzerClient.ts [hunk 1]
```diff
@@ -106,21 +120,19 @@ export class AnalyzerClient {
...
-          //   max_new_tokens: 2048,
-          // },
+          model: getConfigKaiProviderModel(),
+          parameters: getConfigKaiProviderParameters(),
         },
       },
-      file_log_level: "debug",
-      demo_mode: true,
+      file_log_level: getConfigLogLevel(),
+      demo_mode: Extension.getInstance(this.extContext).isProductionMode ? "false" : "true",
```
Comment (rszwajko) [inline vscode/src/client/analyzerClient.ts:135]: can we make it a hidden config property?  `true | false | devMode (default)`

File: vscode/src/utilities/configuration.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+export async function updateContextLines(value: number): Promise<void> {
+  await updateConfigValue("analysis.contextLines", value, vscode.ConfigurationTarget.Workspace);
+}
+
+export async function updateCodeSnipLimit(value: number): Promise<void> {
+  await updateConfigValue("analysis.codeSnipLimit", value, vscode.ConfigurationTarget.Workspace);
+}
+
+export async function updateUseDefaultRulesets(value: boolean): Promise<void> {
```
Comment (rszwajko) [inline vscode/src/utilities/configuration.ts:108]: In the current naming scheme rule sets are joined using camel case ( `ruleSets`). Spell-checker reports that as an error which adds extra noise.
image

**Review by rszwajko — DISMISSED** [review]
Works for me! 2 minor comments (see below)

**Review by sjd78 — APPROVED** [review]
The config changes look good to me.

I didn't realize the kai rpc server needs a path to the analyzer in the initializeParams.  Good to know!

File: vscode/src/client/analyzerClient.ts [hunk 3]
```diff
@@ -412,39 +430,18 @@ export class AnalyzerClient {
...
-  }
-
-  public getContextLines(): number {
-    return this.config!.get("contextLines") as number;
-  }
-
-  public getCodeSnipLimit(): number {
-    return this.config!.get("codeSnipLimit") as number;
-  }
+  public getRules(): string[] {
```
Comment (ibolton336) [inline vscode/src/client/analyzerClient.ts:433]: Returning an array seems to create an issue [here](https://github.com/konveyor/editor-extensions/pull/126/files#diff-7bf6d907ee218343f5466cc2b00dd02fb1e52c9e918671e641a83fb1a8295d6aL139)

**Review by ibolton336 — CHANGES_REQUESTED** [review]

File: vscode/src/client/analyzerClient.ts [hunk 4]
```diff
@@ -106,21 +130,19 @@ export class AnalyzerClient {
...
       model_provider: {
-        provider: "ChatOpenAI",
+        provider: getConfigKaiProviderName(),
         args: {
-          model: "gpt-3.5-turbo",
-          // parameters: {
-          //   max_new_tokens: 2048,
-          // },
+          model: getConfigKaiProviderModel(),
+          parameters: getConfigKaiProviderParameters(),
```
Comment (ibolton336) [inline vscode/src/client/analyzerClient.ts:141]: Something seems to fail when passing in the parameters argument for me:

```ERROR - 2024-12-03 13:50:01,363 - kai.kai.reactive_codeplanner.task_manager.task_manager - Thread-1 - [task_manager.py:112 - supply_result()] - Encountered errors: ["Completions.create() got an unexpected keyword argument 'parameters'"]```

Comment (sjd78) [reply vscode/src/client/analyzerClient.ts:141]: Ian which kai rpc binary are you using?  This is something I can't test w/o linux binaries.

File: vscode/src/client/analyzerClient.ts (hunk 3 above)
Comment (sjd78) [reply vscode/src/client/analyzerClient.ts:433]: @djzager - The `initilizeParams.analyzer_lsp_rules_path` where this function is being called...can that be an array of strings on the `initialize` request or does the array need to be joined before the request?

Comment (djzager) [reply vscode/src/client/analyzerClient.ts:433]: Looks like I had gotten confused. The analyzer itself can take multiple paths, but the kai-analyzer-rpc (and kai-rpc-server by extension) only takes a single path.

I'll fix this now, once I have a tracking issue in kai for the future.

File: vscode/src/client/analyzerClient.ts (hunk 4 above)
Comment (ibolton336) [reply vscode/src/client/analyzerClient.ts]: Yeah that is the one @sjd78. Using an m1 mac luckily.

File: vscode/package.json [hunk 5]
```diff
@@ -403,33 +410,62 @@
...
+          "description": "URL for the Kai solution server",
+          "scope": "window"
+        },
+        "konveyor.kai.providerArgs": {
+          "type": "object",
+          "default": {
+            "model_id": "meta-llama/llama-3-1-70b-instruct",
+            "parameters": {
+              "max_new_tokens": 2048
+            }
```
Comment (ibolton336) [inline vscode/package.json:465]: During my tests I need this to be:
```suggestion
            "model": "gpt-3.5-turbo"
       
```
Else I get errors like:

```Encountered errors: ["Completions.create() got an unexpected keyword argument 'model-id'"]```

**Review by ibolton336 — APPROVED** [review]

File: vscode/package.json (hunk 5 above)
Comment (sjd78) [reply vscode/package.json:465]: Hopefully these model configs don’t become a moving target. I do get the feeling it’ll end up a generic JSON “clob” to handle variations between models on the far back end.

**Pull request merged**